File: Python/1764_Form_Array_By_Concat.py
class Solution:
    def canChoose(self, groups: List[List[int]], nums: List[int]) -> bool:
        j = 0
        M = len(nums)
        
        for grp in groups:
            N = len(grp)
            i = 0
            past_j = j
            while i < N and j < M:
                if grp[i] == nums[j]:
                    i += 1
                else:
                    i = 0
                    j = past_j
                    past_j += 1
                j += 1
            
            if i < N:
                return False
        return True

File: CPlusPlus/1764_Form_Array_By_Concat.cpp
class Solution {
public:
    bool canChoose(vector<vector<int>>& groups, vector<int>& nums) {
        int j = 0, M = nums.size();

        for (auto& grp : groups) {
            int i = 0, N = grp.size();
            int prev_j = j;
            while (i < N && j < M) {
                if (grp[i] == nums[j]) {
                    i += 1;
                } else {
                    i = 0;
                    j = prev_j;
                    prev_j += 1;
                }
                j += 1;
            }
            if (i < N) {
                return false;
            }
        }
        return true;
    }
};
